{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Import standard libraries\n",
    "import os\n",
    "import sys\n",
    "from tqdm import tqdm, trange\n",
    "from glob import glob\n",
    "import pandas as pd\n",
    "import random\n",
    "import logging\n",
    "import argparse\n",
    "\n",
    "## Import custom libraries\n",
    "sys.path.append('/scratch/shared_data_new/metagenomics_graph/model_training/scripts')\n",
    "from utils import get_logger, read_tsv_file, set_seed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Args():\n",
    "    pass\n",
    "\n",
    "args = Args()\n",
    "\n",
    "\n",
    "args.existing_KG_nodes = '/scratch/shared_data_new/metagenomics_graph/KG/KG_nodes_v6.tsv'\n",
    "args.existing_KG_edges = '/scratch/shared_data_new/metagenomics_graph/KG/KG_edges_v6.tsv'\n",
    "args.micobial_hierarchy_dir = '/scratch/shared_data_new/metagenomics_graph/Micobial_hierarchy'\n",
    "args.output_dir = '/scratch/shared_data_new/metagenomics_graph/model_training/data'\n",
    "args.random_seed = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_specific_parent(microbe_id, ranks, microbe_info, find_parent):\n",
    "    \"\"\"\n",
    "    Find the specific parent of a microbe\n",
    "    \"\"\"\n",
    "    \n",
    "    def _recursive_find_parent(microbe_id, ranks, microbe_info, find_parent, res, count=0):\n",
    "        \"\"\"\n",
    "        iterate all the way up to the top of the hierarchy and find all parents belong to the specific ranks\n",
    "        \"\"\"\n",
    "        if count == 8:\n",
    "            return res\n",
    "        \n",
    "        count += 1\n",
    "        if microbe_id not in find_parent:\n",
    "            return res\n",
    "        elif len(microbe_info[microbe_id][0].intersection(set(ranks))) > 0:\n",
    "            res += [microbe_id]\n",
    "            return _recursive_find_parent(find_parent[microbe_id], ranks, microbe_info, find_parent, res, count)\n",
    "        else:\n",
    "            return _recursive_find_parent(find_parent[microbe_id], ranks, microbe_info, find_parent, res, count)\n",
    "    \n",
    "    res = _recursive_find_parent(microbe_id, ranks, microbe_info, find_parent, [], 0)\n",
    "    if len(res) == 0:\n",
    "        return [microbe_id]\n",
    "    else:\n",
    "        return res + [microbe_id]\n",
    "\n",
    "def check_parent(microbe_id, excluded_microbe_ids, find_parent, count=0):\n",
    "    \"\"\"\n",
    "    Check if the parent of a microbe is in the excluded microbe ids\n",
    "    \"\"\"\n",
    "    if count == 8:\n",
    "        return True\n",
    "    \n",
    "    if microbe_id in excluded_microbe_ids:\n",
    "        return False\n",
    "    \n",
    "    if microbe_id not in find_parent:\n",
    "        return True\n",
    "    else:\n",
    "        count += 1\n",
    "        return check_parent(find_parent[microbe_id], excluded_microbe_ids, find_parent, count)\n",
    "\n",
    "def extract_info(sysnonyms, microbe_info_dict, find_parent):\n",
    "    \"\"\"\n",
    "    Get the rank and parent of a microbe based on its id\n",
    "    \"\"\"\n",
    "    if isinstance(sysnonyms, str):\n",
    "        sysnonyms = eval(sysnonyms)\n",
    "\n",
    "    rank_list = []\n",
    "    taxon_id_list = []\n",
    "    parent_list = []\n",
    "    for synonym in sysnonyms:\n",
    "        prefix = synonym.split(':')[0]\n",
    "        if prefix not in ['NCBI', 'GTDB']:\n",
    "            continue\n",
    "        if synonym not in microbe_info_dict:\n",
    "            continue\n",
    "        rank_list += list(microbe_info_dict[synonym][0])\n",
    "        taxon_id_list += list(microbe_info_dict[synonym][1])\n",
    "        parent_list.append(find_parent[synonym])\n",
    "\n",
    "    parent_rank_list = []\n",
    "    parent_taxon_id_list = []\n",
    "    for parent in parent_list:\n",
    "        parent_rank_list += list(microbe_info_dict[parent][0])\n",
    "        parent_taxon_id_list += list(microbe_info_dict[parent][1])\n",
    "    \n",
    "    sysnonyms = str(sysnonyms)\n",
    "    rank_list = str(rank_list)\n",
    "    taxon_id_list = str(taxon_id_list)\n",
    "    parent_list = str(parent_list)\n",
    "    parent_rank_list = str(parent_rank_list)\n",
    "    parent_taxon_id_list = str(parent_taxon_id_list)\n",
    "    \n",
    "    return sysnonyms, rank_list, taxon_id_list, parent_list, parent_rank_list, parent_taxon_id_list\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-01-10 11:52:21  [INFO]  Loading existing knowledge graph nodes and edges...\n"
     ]
    }
   ],
   "source": [
    "# Create a logger object\n",
    "logger = get_logger()\n",
    "logger.setLevel(logging.DEBUG)\n",
    "\n",
    "# set the seed for reproducibility\n",
    "set_seed(args.random_seed)\n",
    "\n",
    "# Load existing knowledge graph nodes and edges\n",
    "logger.info(\"Loading existing knowledge graph nodes and edges...\")\n",
    "KG_nodes_data = read_tsv_file(args.existing_KG_nodes)\n",
    "KG_nodes_header = KG_nodes_data[0]\n",
    "KG_edges_data = read_tsv_file(args.existing_KG_edges)\n",
    "KG_edges_header = KG_edges_data[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Reading microbial hierarchy files:   0%|          | 0/4 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-01-10 11:53:42  [INFO]  Reading /scratch/shared_data_new/metagenomics_graph/Micobial_hierarchy/viruses_hierarchy.tsv...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Reading microbial hierarchy files:  25%|██▌       | 1/4 [00:00<00:02,  1.11it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-01-10 11:53:43  [INFO]  Reading /scratch/shared_data_new/metagenomics_graph/Micobial_hierarchy/fungi_hierarchy.tsv...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Reading microbial hierarchy files:  50%|█████     | 2/4 [00:10<00:11,  5.95s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-01-10 11:53:52  [INFO]  Reading /scratch/shared_data_new/metagenomics_graph/Micobial_hierarchy/bacteria_hierarchy.tsv...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Reading microbial hierarchy files:  75%|███████▌  | 3/4 [00:12<00:04,  4.16s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-01-10 11:53:54  [INFO]  Reading /scratch/shared_data_new/metagenomics_graph/Micobial_hierarchy/archaea_hierarchy.tsv...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Reading microbial hierarchy files: 100%|██████████| 4/4 [00:12<00:00,  3.13s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-01-10 11:53:54  [INFO]  Extracting MKG for GNN model training...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Filtering out the microbe-disease edges: 100%|██████████| 46218489/46218489 [00:58<00:00, 790896.98it/s]\n"
     ]
    }
   ],
   "source": [
    "# Load microbial hierarchy data\n",
    "find_parent = dict()\n",
    "find_child = dict()\n",
    "microbe_info_dict = dict()\n",
    "microbial_hierarchy_files = os.listdir(args.micobial_hierarchy_dir)\n",
    "for microbial_hierarchy_file in tqdm(microbial_hierarchy_files, desc='Reading microbial hierarchy files'):\n",
    "    if microbial_hierarchy_file.split('_')[0] in ['viruses', 'fungi']:\n",
    "        prefix = 'NCBI:'\n",
    "    else:\n",
    "        prefix = 'GTDB:'\n",
    "    microbial_hierarchy_file_path = os.path.join(args.micobial_hierarchy_dir, microbial_hierarchy_file)\n",
    "    logger.info(\"Reading {}...\".format(microbial_hierarchy_file_path))\n",
    "    temp_data = read_tsv_file(microbial_hierarchy_file_path)\n",
    "    for row in temp_data[1:]:\n",
    "        row[0] = row[0].replace('GB_', '').replace('RS_', '')\n",
    "        row[3] = row[3].replace('GB_', '').replace('RS_', '')\n",
    "        if f\"{prefix}{row[0]}\" not in microbe_info_dict:\n",
    "            microbe_info_dict[f\"{prefix}{row[0]}\"] = [set(), set()]\n",
    "        microbe_info_dict[f\"{prefix}{row[0]}\"][0].add(row[1])\n",
    "        microbe_info_dict[f\"{prefix}{row[0]}\"][1].add(row[2])              \n",
    "        if f\"{prefix}{row[3]}\" not in microbe_info_dict:\n",
    "            microbe_info_dict[f\"{prefix}{row[3]}\"] = [set(), set()]\n",
    "        microbe_info_dict[f\"{prefix}{row[3]}\"][0].add(row[4])\n",
    "        microbe_info_dict[f\"{prefix}{row[3]}\"][1].add(row[5])\n",
    "        if f\"{prefix}{row[3]}\" != f\"{prefix}{row[0]}\":\n",
    "            find_parent[f\"{prefix}{row[3]}\"] = f\"{prefix}{row[0]}\"\n",
    "            if f\"{prefix}{row[0]}\" not in find_child:\n",
    "                find_child[f\"{prefix}{row[0]}\"] = set()\n",
    "            find_child[f\"{prefix}{row[0]}\"].add(f\"{prefix}{row[3]}\")\n",
    "\n",
    "## Extract data from MKG for GNN model training \n",
    "logger.info(\"Extracting MKG for GNN model training...\")\n",
    "node_info = pd.DataFrame([(row[KG_nodes_header.index('node_id')], row[KG_nodes_header.index('node_type')], \\\n",
    "                            row[KG_nodes_header.index('all_names')], row[KG_nodes_header.index('synonyms')], row[KG_nodes_header.index('is_pathogen')]) \\\n",
    "                            for row in KG_nodes_data[1:]])\n",
    "node_info.columns = ['node_id', 'node_type', 'all_names', 'synonyms', 'is_pathogen']\n",
    "edge_info = pd.DataFrame([(row[KG_edges_header.index('source_node')], row[KG_edges_header.index('target_node')], row[KG_edges_header.index('predicate')]) for row in KG_edges_data[1:]])\n",
    "edge_info.columns = ['source_node', 'target_node', 'predicate']\n",
    "\n",
    "# Filter out the microbe-disease edges\n",
    "edge_info_temp = []\n",
    "filtered_type = ['Microbe', 'Disease']\n",
    "for row in tqdm(edge_info.to_numpy(), desc=\"Filtering out the microbe-disease edges\"):\n",
    "    source_node, target_node, predicate = row\n",
    "    source_type = source_node.split(':')[0]\n",
    "    target_type = target_node.split(':')[0]\n",
    "    if source_type == target_type:\n",
    "        edge_info_temp.append(row)\n",
    "    else:\n",
    "        if source_type in filtered_type and target_type in filtered_type:\n",
    "            continue\n",
    "        else:\n",
    "            edge_info_temp.append(row)\n",
    "edge_info = pd.DataFrame(edge_info_temp)\n",
    "edge_info.columns = ['source_node', 'target_node', 'predicate']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Covert node and edge to index\n",
    "# save node to index mapping\n",
    "all_nodes = node_info['node_id'].to_list()\n",
    "random.shuffle(all_nodes)\n",
    "node_to_index = pd.DataFrame(all_nodes).reset_index()[[0, 'index']]\n",
    "node_to_index.columns = ['node_id', 'node_index']\n",
    "node_to_index.to_csv(os.path.join(args.output_dir, 'node_to_index.tsv'), sep='\\t', index=False)\n",
    "node_info.to_csv(os.path.join(args.output_dir, 'node_info.tsv'), sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save pathogen info\n",
    "pathogen_info = node_info.loc[node_info['is_pathogen']=='True', :].reset_index(drop=True)\n",
    "# find pathogen rank and its parent info\n",
    "part1 = pathogen_info[['node_id', 'node_type', 'all_names', 'is_pathogen']]\n",
    "part2 = pd.DataFrame(pathogen_info['synonyms'].apply(lambda x: extract_info(x, microbe_info_dict, find_parent)).to_list(), columns=['synonyms', 'rank_list', 'taxon_id_list', 'parent_list', 'parent_rank_list', 'parent_taxon_id_list'])\n",
    "pathogen_info = pd.concat([part1, part2], axis=1)\n",
    "pathogen_info = pathogen_info.loc[pathogen_info['synonyms'].str.contains('GTDB'),:].reset_index(drop=True)\n",
    "pathogen_info = pathogen_info.loc[pathogen_info['rank_list'].str.contains('species|strain'),:].reset_index(drop=True)\n",
    "pathogen_info.to_csv(os.path.join(args.output_dir, 'pathogen_info.tsv'), sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 708/708 [00:00<00:00, 28418.81it/s]\n",
      "100%|██████████| 951782/951782 [00:03<00:00, 304568.22it/s] \n"
     ]
    }
   ],
   "source": [
    "# find all potential pathogens (all strains under the known species-level pathogens that have at least one pathogen or themself are known as pathogens )\n",
    "species1 = set(pathogen_info.loc[pathogen_info['rank_list'].str.contains('strain'),'parent_list'].to_list())\n",
    "species2 = set(pathogen_info.loc[pathogen_info['rank_list'].str.contains('species'),'synonyms'].to_list())\n",
    "merged_species = species1 | species2\n",
    "all_potential_pathogenic_strain = set()\n",
    "for species in tqdm(merged_species):\n",
    "    species_synonyms_list = eval(species)\n",
    "    for species_synonyms in species_synonyms_list:\n",
    "        all_potential_pathogenic_strain |= find_child[species_synonyms]\n",
    "microbe_nodes = node_info.loc[node_info['node_id'].str.contains('Microbe'), :].reset_index(drop=True)\n",
    "all_gtdb_id_to_microbe_id = dict()\n",
    "for row in tqdm(microbe_nodes.to_numpy()):\n",
    "    node_id, node_type, all_names, synonyms, is_pathogen = row\n",
    "    if 'GTDB:' not in synonyms:\n",
    "        continue\n",
    "    all_gtdb_id_to_microbe_id\n",
    "    synonyms = eval(synonyms)\n",
    "    for synonym in synonyms:\n",
    "        if 'GTDB:' in synonym:\n",
    "            all_gtdb_id_to_microbe_id[synonym] = (synonym, node_id, node_type, all_names, synonyms, is_pathogen)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_gtdb_id_to_microbe_id = dict()\n",
    "for row in tqdm(node_info.to_numpy()):\n",
    "    node_id, node_type, all_names, synonyms, is_pathogen = row\n",
    "    if 'GTDB:' not in synonyms:\n",
    "        continue\n",
    "    all_gtdb_id_to_microbe_id\n",
    "    synonyms = eval(synonyms)\n",
    "    for synonym in synonyms:\n",
    "        if 'GTDB:' in synonym:\n",
    "            all_gtdb_id_to_microbe_id[synonym] = (synonym, node_id, node_type, all_names, is_pathogen, synonyms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/951782 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 951782/951782 [00:03<00:00, 254570.53it/s] \n"
     ]
    }
   ],
   "source": [
    "gtdbid_df = pd.DataFrame(all_gtdb_id_to_microbe_id.values(), columns=['gtdb_id', 'node_id', 'node_type', 'all_names', 'is_pathogen', 'synonyms'])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "708"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(merged_species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gtdb_id</th>\n",
       "      <th>node_id</th>\n",
       "      <th>node_type</th>\n",
       "      <th>all_names</th>\n",
       "      <th>synonyms</th>\n",
       "      <th>is_pathogen</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GTDB:Bacteria</td>\n",
       "      <td>Microbe:431882</td>\n",
       "      <td>biolink:OrganismTaxon</td>\n",
       "      <td>['Bacteria']</td>\n",
       "      <td>[GTDB:Bacteria]</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GTDB:Pseudomonadota</td>\n",
       "      <td>Microbe:431883</td>\n",
       "      <td>biolink:OrganismTaxon</td>\n",
       "      <td>['Pseudomonadota']</td>\n",
       "      <td>[GTDB:Pseudomonadota]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GTDB:Gammaproteobacteria</td>\n",
       "      <td>Microbe:431884</td>\n",
       "      <td>biolink:OrganismTaxon</td>\n",
       "      <td>['Gammaproteobacteria']</td>\n",
       "      <td>[GTDB:Gammaproteobacteria]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GTDB:Enterobacterales</td>\n",
       "      <td>Microbe:431885</td>\n",
       "      <td>biolink:OrganismTaxon</td>\n",
       "      <td>['Enterobacterales']</td>\n",
       "      <td>[GTDB:Enterobacterales]</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GTDB:Enterobacteriaceae</td>\n",
       "      <td>Microbe:431886</td>\n",
       "      <td>biolink:OrganismTaxon</td>\n",
       "      <td>['Enterobacteriaceae']</td>\n",
       "      <td>[GTDB:Enterobacteriaceae]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>510674</th>\n",
       "      <td>GTDB:GCA_000180295.1</td>\n",
       "      <td>Microbe:944250</td>\n",
       "      <td>biolink:OrganismTaxon</td>\n",
       "      <td>['Salmonella enterica subsp. enterica serovar ...</td>\n",
       "      <td>[BVBRC:gn_496068.3, GTDB:GCA_000180295.1]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>510675</th>\n",
       "      <td>GTDB:GCA_000268065.1</td>\n",
       "      <td>Microbe:944788</td>\n",
       "      <td>biolink:OrganismTaxon</td>\n",
       "      <td>['Shigella flexneri 1235-66']</td>\n",
       "      <td>[BVBRC:gn_766154.3, GTDB:GCA_000268065.1]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>510676</th>\n",
       "      <td>GTDB:GCA_000017125.1</td>\n",
       "      <td>Microbe:945749</td>\n",
       "      <td>biolink:OrganismTaxon</td>\n",
       "      <td>['Staphylococcus aureus subsp. aureus JH1']</td>\n",
       "      <td>[BVBRC:gn_359787.11, GTDB:GCA_000017125.1]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>510677</th>\n",
       "      <td>GTDB:GCA_002249745.1</td>\n",
       "      <td>Microbe:945750</td>\n",
       "      <td>biolink:OrganismTaxon</td>\n",
       "      <td>['Streptococcus pneumoniae B1598']</td>\n",
       "      <td>[BVBRC:gn_1449970.3, GTDB:GCA_002249745.1]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>510678</th>\n",
       "      <td>GTDB:GCA_000305735.1</td>\n",
       "      <td>Microbe:951782</td>\n",
       "      <td>biolink:OrganismTaxon</td>\n",
       "      <td>['Vibrio sp. HENC-02']</td>\n",
       "      <td>[GTDB:GCA_000305735.1, BVBRC:gn_992011.3]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>510679 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         gtdb_id         node_id              node_type  \\\n",
       "0                  GTDB:Bacteria  Microbe:431882  biolink:OrganismTaxon   \n",
       "1            GTDB:Pseudomonadota  Microbe:431883  biolink:OrganismTaxon   \n",
       "2       GTDB:Gammaproteobacteria  Microbe:431884  biolink:OrganismTaxon   \n",
       "3          GTDB:Enterobacterales  Microbe:431885  biolink:OrganismTaxon   \n",
       "4        GTDB:Enterobacteriaceae  Microbe:431886  biolink:OrganismTaxon   \n",
       "...                          ...             ...                    ...   \n",
       "510674      GTDB:GCA_000180295.1  Microbe:944250  biolink:OrganismTaxon   \n",
       "510675      GTDB:GCA_000268065.1  Microbe:944788  biolink:OrganismTaxon   \n",
       "510676      GTDB:GCA_000017125.1  Microbe:945749  biolink:OrganismTaxon   \n",
       "510677      GTDB:GCA_002249745.1  Microbe:945750  biolink:OrganismTaxon   \n",
       "510678      GTDB:GCA_000305735.1  Microbe:951782  biolink:OrganismTaxon   \n",
       "\n",
       "                                                all_names  \\\n",
       "0                                            ['Bacteria']   \n",
       "1                                      ['Pseudomonadota']   \n",
       "2                                 ['Gammaproteobacteria']   \n",
       "3                                    ['Enterobacterales']   \n",
       "4                                  ['Enterobacteriaceae']   \n",
       "...                                                   ...   \n",
       "510674  ['Salmonella enterica subsp. enterica serovar ...   \n",
       "510675                      ['Shigella flexneri 1235-66']   \n",
       "510676        ['Staphylococcus aureus subsp. aureus JH1']   \n",
       "510677                 ['Streptococcus pneumoniae B1598']   \n",
       "510678                             ['Vibrio sp. HENC-02']   \n",
       "\n",
       "                                          synonyms is_pathogen  \n",
       "0                                  [GTDB:Bacteria]       False  \n",
       "1                            [GTDB:Pseudomonadota]        True  \n",
       "2                       [GTDB:Gammaproteobacteria]        True  \n",
       "3                          [GTDB:Enterobacterales]       False  \n",
       "4                        [GTDB:Enterobacteriaceae]        True  \n",
       "...                                            ...         ...  \n",
       "510674   [BVBRC:gn_496068.3, GTDB:GCA_000180295.1]        True  \n",
       "510675   [BVBRC:gn_766154.3, GTDB:GCA_000268065.1]        True  \n",
       "510676  [BVBRC:gn_359787.11, GTDB:GCA_000017125.1]        True  \n",
       "510677  [BVBRC:gn_1449970.3, GTDB:GCA_002249745.1]        True  \n",
       "510678   [GTDB:GCA_000305735.1, BVBRC:gn_992011.3]        True  \n",
       "\n",
       "[510679 rows x 6 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtdbid_df = pd.DataFrame(all_gtdb_id_to_microbe_id.values(), columns=['gtdb_id', 'node_id', 'node_type', 'all_names', 'synonyms', 'is_pathogen'])\n",
    "gtdbid_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "402732"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(gtdbid_df['node_id'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtdbid_df = pd.DataFrame(all_gtdb_id_to_microbe_id.values(), columns=['gtdb_id', 'node_id', 'node_type', 'all_names', 'synonyms', 'is_pathogen'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gtdb_id</th>\n",
       "      <th>node_id</th>\n",
       "      <th>node_type</th>\n",
       "      <th>all_names</th>\n",
       "      <th>synonyms</th>\n",
       "      <th>is_pathogen</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GTDB:Bacteria</td>\n",
       "      <td>Microbe:431882</td>\n",
       "      <td>biolink:OrganismTaxon</td>\n",
       "      <td>['Bacteria']</td>\n",
       "      <td>[GTDB:Bacteria]</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GTDB:Pseudomonadota</td>\n",
       "      <td>Microbe:431883</td>\n",
       "      <td>biolink:OrganismTaxon</td>\n",
       "      <td>['Pseudomonadota']</td>\n",
       "      <td>[GTDB:Pseudomonadota]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GTDB:Gammaproteobacteria</td>\n",
       "      <td>Microbe:431884</td>\n",
       "      <td>biolink:OrganismTaxon</td>\n",
       "      <td>['Gammaproteobacteria']</td>\n",
       "      <td>[GTDB:Gammaproteobacteria]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GTDB:Enterobacterales</td>\n",
       "      <td>Microbe:431885</td>\n",
       "      <td>biolink:OrganismTaxon</td>\n",
       "      <td>['Enterobacterales']</td>\n",
       "      <td>[GTDB:Enterobacterales]</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GTDB:Enterobacteriaceae</td>\n",
       "      <td>Microbe:431886</td>\n",
       "      <td>biolink:OrganismTaxon</td>\n",
       "      <td>['Enterobacteriaceae']</td>\n",
       "      <td>[GTDB:Enterobacteriaceae]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>510674</th>\n",
       "      <td>GTDB:GCA_000180295.1</td>\n",
       "      <td>Microbe:944250</td>\n",
       "      <td>biolink:OrganismTaxon</td>\n",
       "      <td>['Salmonella enterica subsp. enterica serovar ...</td>\n",
       "      <td>[BVBRC:gn_496068.3, GTDB:GCA_000180295.1]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>510675</th>\n",
       "      <td>GTDB:GCA_000268065.1</td>\n",
       "      <td>Microbe:944788</td>\n",
       "      <td>biolink:OrganismTaxon</td>\n",
       "      <td>['Shigella flexneri 1235-66']</td>\n",
       "      <td>[BVBRC:gn_766154.3, GTDB:GCA_000268065.1]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>510676</th>\n",
       "      <td>GTDB:GCA_000017125.1</td>\n",
       "      <td>Microbe:945749</td>\n",
       "      <td>biolink:OrganismTaxon</td>\n",
       "      <td>['Staphylococcus aureus subsp. aureus JH1']</td>\n",
       "      <td>[BVBRC:gn_359787.11, GTDB:GCA_000017125.1]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>510677</th>\n",
       "      <td>GTDB:GCA_002249745.1</td>\n",
       "      <td>Microbe:945750</td>\n",
       "      <td>biolink:OrganismTaxon</td>\n",
       "      <td>['Streptococcus pneumoniae B1598']</td>\n",
       "      <td>[BVBRC:gn_1449970.3, GTDB:GCA_002249745.1]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>510678</th>\n",
       "      <td>GTDB:GCA_000305735.1</td>\n",
       "      <td>Microbe:951782</td>\n",
       "      <td>biolink:OrganismTaxon</td>\n",
       "      <td>['Vibrio sp. HENC-02']</td>\n",
       "      <td>[GTDB:GCA_000305735.1, BVBRC:gn_992011.3]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>510679 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         gtdb_id         node_id              node_type  \\\n",
       "0                  GTDB:Bacteria  Microbe:431882  biolink:OrganismTaxon   \n",
       "1            GTDB:Pseudomonadota  Microbe:431883  biolink:OrganismTaxon   \n",
       "2       GTDB:Gammaproteobacteria  Microbe:431884  biolink:OrganismTaxon   \n",
       "3          GTDB:Enterobacterales  Microbe:431885  biolink:OrganismTaxon   \n",
       "4        GTDB:Enterobacteriaceae  Microbe:431886  biolink:OrganismTaxon   \n",
       "...                          ...             ...                    ...   \n",
       "510674      GTDB:GCA_000180295.1  Microbe:944250  biolink:OrganismTaxon   \n",
       "510675      GTDB:GCA_000268065.1  Microbe:944788  biolink:OrganismTaxon   \n",
       "510676      GTDB:GCA_000017125.1  Microbe:945749  biolink:OrganismTaxon   \n",
       "510677      GTDB:GCA_002249745.1  Microbe:945750  biolink:OrganismTaxon   \n",
       "510678      GTDB:GCA_000305735.1  Microbe:951782  biolink:OrganismTaxon   \n",
       "\n",
       "                                                all_names  \\\n",
       "0                                            ['Bacteria']   \n",
       "1                                      ['Pseudomonadota']   \n",
       "2                                 ['Gammaproteobacteria']   \n",
       "3                                    ['Enterobacterales']   \n",
       "4                                  ['Enterobacteriaceae']   \n",
       "...                                                   ...   \n",
       "510674  ['Salmonella enterica subsp. enterica serovar ...   \n",
       "510675                      ['Shigella flexneri 1235-66']   \n",
       "510676        ['Staphylococcus aureus subsp. aureus JH1']   \n",
       "510677                 ['Streptococcus pneumoniae B1598']   \n",
       "510678                             ['Vibrio sp. HENC-02']   \n",
       "\n",
       "                                          synonyms is_pathogen  \n",
       "0                                  [GTDB:Bacteria]       False  \n",
       "1                            [GTDB:Pseudomonadota]        True  \n",
       "2                       [GTDB:Gammaproteobacteria]        True  \n",
       "3                          [GTDB:Enterobacterales]       False  \n",
       "4                        [GTDB:Enterobacteriaceae]        True  \n",
       "...                                            ...         ...  \n",
       "510674   [BVBRC:gn_496068.3, GTDB:GCA_000180295.1]        True  \n",
       "510675   [BVBRC:gn_766154.3, GTDB:GCA_000268065.1]        True  \n",
       "510676  [BVBRC:gn_359787.11, GTDB:GCA_000017125.1]        True  \n",
       "510677  [BVBRC:gn_1449970.3, GTDB:GCA_002249745.1]        True  \n",
       "510678   [GTDB:GCA_000305735.1, BVBRC:gn_992011.3]        True  \n",
       "\n",
       "[510679 rows x 6 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtdbid_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_potential_pathogenic_strain_df = gtdbid_df.loc[gtdbid_df['gtdb_id'].isin(all_potential_pathogenic_strain),:].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_potential_pathogenic_strain_df.to_csv(os.path.join(args.output_dir, 'all_potential_pathogenic_strain.tsv'), sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{\"['GTDB:Achromobacter insolitus']\",\n",
       " \"['GTDB:Achromobacter ruhlandii']\",\n",
       " \"['GTDB:Achromobacter spanius']\",\n",
       " \"['GTDB:Achromobacter xylosoxidans']\",\n",
       " \"['GTDB:Acidaminococcus intestini']\",\n",
       " \"['GTDB:Acinetobacter baumannii']\",\n",
       " \"['GTDB:Acinetobacter calcoaceticus']\",\n",
       " \"['GTDB:Acinetobacter fasciculus']\",\n",
       " \"['GTDB:Acinetobacter haemolyticus']\",\n",
       " \"['GTDB:Acinetobacter johnsonii']\",\n",
       " \"['GTDB:Acinetobacter junii']\",\n",
       " \"['GTDB:Acinetobacter radioresistens']\",\n",
       " \"['GTDB:Acinetobacter schindleri']\",\n",
       " \"['GTDB:Actinomyces oris']\",\n",
       " \"['GTDB:Actinotignum schaalii']\",\n",
       " \"['GTDB:Aerococcus christensenii']\",\n",
       " \"['GTDB:Aerococcus sanguinicola']\",\n",
       " \"['GTDB:Aerococcus urinae']\",\n",
       " \"['GTDB:Aerococcus urinaehominis']\",\n",
       " \"['GTDB:Aeromonas caviae']\",\n",
       " \"['GTDB:Aeromonas enteropelogenes']\",\n",
       " \"['GTDB:Aeromonas hydrophila']\",\n",
       " \"['GTDB:Aggregatibacter actinomycetemcomitans']\",\n",
       " \"['GTDB:Aggregatibacter actinomycetemcomitans_A']\",\n",
       " \"['GTDB:Aggregatibacter aphrophilus']\",\n",
       " \"['GTDB:Aliarcobacter butzleri']\",\n",
       " \"['GTDB:Alistipes putredinis']\",\n",
       " \"['GTDB:Alloiococcus otitis']\",\n",
       " \"['GTDB:Anaplasma phagocytophilum']\",\n",
       " \"['GTDB:Bacillus_A anthracis']\",\n",
       " \"['GTDB:Bacillus_A cereus']\",\n",
       " \"['GTDB:Bacillus_A cytotoxicus']\",\n",
       " \"['GTDB:Bacillus_A mycoides']\",\n",
       " \"['GTDB:Bacillus_A nitratireducens']\",\n",
       " \"['GTDB:Bacillus_A paranthracis']\",\n",
       " \"['GTDB:Bacillus_A thuringiensis_S']\",\n",
       " \"['GTDB:Bacillus_A tropicus']\",\n",
       " \"['GTDB:Bacteroides caccae']\",\n",
       " \"['GTDB:Bacteroides fragilis']\",\n",
       " \"['GTDB:Bartonella ancashensis']\",\n",
       " \"['GTDB:Bartonella bacilliformis']\",\n",
       " \"['GTDB:Bartonella grahamii']\",\n",
       " \"['GTDB:Bartonella henselae']\",\n",
       " \"['GTDB:Bartonella quintana']\",\n",
       " \"['GTDB:Bergeyella_A cardium']\",\n",
       " \"['GTDB:Bifidobacterium animalis']\",\n",
       " \"['GTDB:Bifidobacterium leopoldii']\",\n",
       " \"['GTDB:Bifidobacterium piotii']\",\n",
       " \"['GTDB:Bifidobacterium vaginale']\",\n",
       " \"['GTDB:Bifidobacterium vaginale_C']\",\n",
       " \"['GTDB:Bifidobacterium vaginale_E']\",\n",
       " \"['GTDB:Bordetella hinzii']\",\n",
       " \"['GTDB:Bordetella holmesii']\",\n",
       " \"['GTDB:Bordetella pertussis']\",\n",
       " \"['GTDB:Bordetella_A petrii']\",\n",
       " \"['GTDB:Bordetella_C bronchialis']\",\n",
       " \"['GTDB:Bordetella_C flabilis']\",\n",
       " \"['GTDB:Bordetella_C sp001704295']\",\n",
       " \"['GTDB:Borrelia coriaceae']\",\n",
       " \"['GTDB:Borrelia hermsii']\",\n",
       " \"['GTDB:Borrelia miyamotoi']\",\n",
       " \"['GTDB:Borrelia recurrentis']\",\n",
       " \"['GTDB:Borrelia turicatae']\",\n",
       " \"['GTDB:Borreliella afzelii']\",\n",
       " \"['GTDB:Borreliella bissettiae']\",\n",
       " \"['GTDB:Borreliella burgdorferi']\",\n",
       " \"['GTDB:Borreliella garinii']\",\n",
       " \"['GTDB:Borreliella mayonii']\",\n",
       " \"['GTDB:Brachyspira pilosicoli']\",\n",
       " \"['GTDB:Bradyrhizobium sp003020075']\",\n",
       " \"['GTDB:Brucella melitensis']\",\n",
       " \"['GTDB:Burkholderia ambifaria']\",\n",
       " \"['GTDB:Burkholderia cenocepacia']\",\n",
       " \"['GTDB:Burkholderia glumae']\",\n",
       " \"['GTDB:Burkholderia mallei']\",\n",
       " \"['GTDB:Burkholderia multivorans']\",\n",
       " \"['GTDB:Burkholderia orbicola']\",\n",
       " \"['GTDB:Burkholderia stabilis']\",\n",
       " \"['GTDB:Campylobacter fetus']\",\n",
       " \"['GTDB:Campylobacter_A concisus_A']\",\n",
       " \"['GTDB:Campylobacter_A concisus_AB']\",\n",
       " \"['GTDB:Campylobacter_A concisus_R']\",\n",
       " \"['GTDB:Campylobacter_A curvus']\",\n",
       " \"['GTDB:Campylobacter_A rectus']\",\n",
       " \"['GTDB:Campylobacter_A showae']\",\n",
       " \"['GTDB:Campylobacter_B ureolyticus']\",\n",
       " \"['GTDB:Campylobacter_D coli']\",\n",
       " \"['GTDB:Campylobacter_D jejuni']\",\n",
       " \"['GTDB:Campylobacter_D upsaliensis']\",\n",
       " \"['GTDB:Capnocytophaga canimorsus']\",\n",
       " \"['GTDB:Capnocytophaga gingivalis']\",\n",
       " \"['GTDB:Capnocytophaga ochracea']\",\n",
       " \"['GTDB:Capnocytophaga sputigena']\",\n",
       " \"['GTDB:Cedecea neteri_C']\",\n",
       " \"['GTDB:Chlamydia trachomatis']\",\n",
       " \"['GTDB:Chlamydophila pneumoniae']\",\n",
       " \"['GTDB:Chlamydophila psittaci']\",\n",
       " \"['GTDB:Chromobacterium violaceum']\",\n",
       " \"['GTDB:Citrobacter_B koseri']\",\n",
       " \"['GTDB:Clostridioides difficile']\",\n",
       " \"['GTDB:Clostridium botulinum_A']\",\n",
       " \"['GTDB:Clostridium_F botulinum']\",\n",
       " \"['GTDB:Clostridium_G tetani']\",\n",
       " \"['GTDB:Clostridium_H haemolyticum']\",\n",
       " \"['GTDB:Clostridium_Q saccharolyticum_A']\",\n",
       " \"['GTDB:Cohnella cholangitidis']\",\n",
       " \"['GTDB:Corynebacterium amycolatum_A']\",\n",
       " \"['GTDB:Corynebacterium argentoratense']\",\n",
       " \"['GTDB:Corynebacterium aurimucosum_F']\",\n",
       " \"['GTDB:Corynebacterium diphtheriae']\",\n",
       " \"['GTDB:Corynebacterium imitans']\",\n",
       " \"['GTDB:Corynebacterium jeikeium']\",\n",
       " \"['GTDB:Corynebacterium kroppenstedtii']\",\n",
       " \"['GTDB:Corynebacterium macginleyi']\",\n",
       " \"['GTDB:Corynebacterium matruchotii']\",\n",
       " \"['GTDB:Corynebacterium mucifaciens']\",\n",
       " \"['GTDB:Corynebacterium resistens']\",\n",
       " \"['GTDB:Corynebacterium ulcerans']\",\n",
       " \"['GTDB:Corynebacterium urealyticum']\",\n",
       " \"['GTDB:Coxiella burnetii']\",\n",
       " \"['GTDB:Cronobacter condimenti']\",\n",
       " \"['GTDB:Cronobacter dublinensis']\",\n",
       " \"['GTDB:Cronobacter malonaticus']\",\n",
       " \"['GTDB:Cronobacter sakazakii']\",\n",
       " \"['GTDB:Cronobacter turicensis']\",\n",
       " \"['GTDB:Cronobacter universalis']\",\n",
       " \"['GTDB:Cryptobacterium curtum']\",\n",
       " \"['GTDB:Cutibacterium acnes']\",\n",
       " \"['GTDB:Cutibacterium namnetense']\",\n",
       " \"['GTDB:Desulfobulbus oralis']\",\n",
       " \"['GTDB:Edwardsiella tarda']\",\n",
       " \"['GTDB:Eggerthella lenta']\",\n",
       " \"['GTDB:Ehrlichia chaffeensis']\",\n",
       " \"['GTDB:Eikenella corrodens']\",\n",
       " \"['GTDB:Elizabethkingia anophelis']\",\n",
       " \"['GTDB:Elizabethkingia miricola']\",\n",
       " \"['GTDB:Enterobacter cancerogenus']\",\n",
       " \"['GTDB:Enterobacter cloacae']\",\n",
       " \"['GTDB:Enterobacter hormaechei']\",\n",
       " \"['GTDB:Enterobacter hormaechei_A']\",\n",
       " \"['GTDB:Enterobacter ludwigii']\",\n",
       " \"['GTDB:Enterobacter roggenkampii']\",\n",
       " \"['GTDB:Enterococcus faecalis']\",\n",
       " \"['GTDB:Enterococcus_A avium']\",\n",
       " \"['GTDB:Enterococcus_A gilvus']\",\n",
       " \"['GTDB:Enterococcus_A pallens']\",\n",
       " \"['GTDB:Enterococcus_B faecium']\",\n",
       " \"['GTDB:Enterococcus_B lactis']\",\n",
       " \"['GTDB:Enterococcus_D casseliflavus']\",\n",
       " \"['GTDB:Erysipelothrix rhusiopathiae']\",\n",
       " \"['GTDB:Escherichia coli']\",\n",
       " \"['GTDB:Escherichia fergusonii']\",\n",
       " \"['GTDB:Escherichia ruysiae']\",\n",
       " \"['GTDB:Finegoldia magna_H']\",\n",
       " \"['GTDB:Flavobacterium odoratimimum']\",\n",
       " \"['GTDB:Francisella tularensis']\",\n",
       " \"['GTDB:Fusobacterium nucleatum']\",\n",
       " \"['GTDB:Fusobacterium periodonticum']\",\n",
       " \"['GTDB:Fusobacterium_A varium_A']\",\n",
       " \"['GTDB:Fusobacterium_C necrophorum']\",\n",
       " \"['GTDB:Gemella haemolysans_A']\",\n",
       " \"['GTDB:Gemella morbillorum']\",\n",
       " \"['GTDB:Gemella sanguinis']\",\n",
       " \"['GTDB:Glutamicibacter creatinolyticus']\",\n",
       " \"['GTDB:Gordonia bronchialis']\",\n",
       " \"['GTDB:Granulibacter bethesdensis']\",\n",
       " \"['GTDB:Haemophilus aegyptius']\",\n",
       " \"['GTDB:Haemophilus influenzae']\",\n",
       " \"['GTDB:Haemophilus influenzae_D']\",\n",
       " \"['GTDB:Haemophilus influenzae_E']\",\n",
       " \"['GTDB:Haemophilus influenzae_G']\",\n",
       " \"['GTDB:Haemophilus_B ducreyi']\",\n",
       " \"['GTDB:Hafnia alvei']\",\n",
       " \"['GTDB:Helicobacter pylori']\",\n",
       " \"['GTDB:Helicobacter pylori_A']\",\n",
       " \"['GTDB:Helicobacter pylori_BU']\",\n",
       " \"['GTDB:Helicobacter pylori_C']\",\n",
       " \"['GTDB:Helicobacter pylori_CI']\",\n",
       " \"['GTDB:Helicobacter_B fennelliae']\",\n",
       " \"['GTDB:Helicobacter_C cinaedi']\",\n",
       " \"['GTDB:Helicobacter_D canadensis']\",\n",
       " \"['GTDB:Helicobacter_D pullorum']\",\n",
       " \"['GTDB:Helicobacter_E heilmannii']\",\n",
       " \"['GTDB:Helicobacter_E suis']\",\n",
       " \"['GTDB:Inquilinus limosus_E']\",\n",
       " \"['GTDB:Kerstersia gyiorum']\",\n",
       " \"['GTDB:Kingella kingae']\",\n",
       " \"['GTDB:Klebsiella aerogenes']\",\n",
       " \"['GTDB:Klebsiella africana']\",\n",
       " \"['GTDB:Klebsiella huaxiensis']\",\n",
       " \"['GTDB:Klebsiella michiganensis']\",\n",
       " \"['GTDB:Klebsiella pneumoniae']\",\n",
       " \"['GTDB:Klebsiella quasipneumoniae']\",\n",
       " \"['GTDB:Klebsiella terrigena']\",\n",
       " \"['GTDB:Kosakonia cowanii']\",\n",
       " \"['GTDB:Lactococcus garvieae']\",\n",
       " \"['GTDB:Lancefieldella rimae']\",\n",
       " \"['GTDB:Laribacter hongkongensis']\",\n",
       " \"['GTDB:Legionella anisa']\",\n",
       " \"['GTDB:Legionella longbeachae']\",\n",
       " \"['GTDB:Legionella pneumophila']\",\n",
       " \"['GTDB:Legionella sainthelensi']\",\n",
       " \"['GTDB:Leptospira borgpetersenii']\",\n",
       " \"['GTDB:Leptospira interrogans']\",\n",
       " \"['GTDB:Leptospira kmetyi']\",\n",
       " \"['GTDB:Leptospira santarosai']\",\n",
       " \"['GTDB:Listeria monocytogenes']\",\n",
       " \"['GTDB:Listeria monocytogenes_B']\",\n",
       " \"['GTDB:Listeria monocytogenes_C']\",\n",
       " \"['GTDB:Malacoplasma penetrans']\",\n",
       " \"['GTDB:Mammaliicoccus lentus']\",\n",
       " \"['GTDB:Metamycoplasma hominis']\",\n",
       " \"['GTDB:Mobiluncus mulieris']\",\n",
       " \"['GTDB:Moraxella catarrhalis']\",\n",
       " \"['GTDB:Morganella morganii']\",\n",
       " \"['GTDB:Mycobacterium abscessus']\",\n",
       " \"['GTDB:Mycobacterium aichiense']\",\n",
       " \"['GTDB:Mycobacterium algericum']\",\n",
       " \"['GTDB:Mycobacterium alvei']\",\n",
       " \"['GTDB:Mycobacterium aubagnense']\",\n",
       " \"['GTDB:Mycobacterium avium']\",\n",
       " \"['GTDB:Mycobacterium celeriflavum']\",\n",
       " \"['GTDB:Mycobacterium chelonae']\",\n",
       " \"['GTDB:Mycobacterium colombiense']\",\n",
       " \"['GTDB:Mycobacterium doricum']\",\n",
       " \"['GTDB:Mycobacterium duvalii']\",\n",
       " \"['GTDB:Mycobacterium florentinum']\",\n",
       " \"['GTDB:Mycobacterium fortuitum']\",\n",
       " \"['GTDB:Mycobacterium gallinarum']\",\n",
       " \"['GTDB:Mycobacterium grossiae']\",\n",
       " \"['GTDB:Mycobacterium haemophilum']\",\n",
       " \"['GTDB:Mycobacterium heckeshornense']\",\n",
       " \"['GTDB:Mycobacterium insubricum']\",\n",
       " \"['GTDB:Mycobacterium intracellulare']\",\n",
       " \"['GTDB:Mycobacterium kansasii']\",\n",
       " \"['GTDB:Mycobacterium koreense']\",\n",
       " \"['GTDB:Mycobacterium leprae']\",\n",
       " \"['GTDB:Mycobacterium lepromatosis']\",\n",
       " \"['GTDB:Mycobacterium marinum']\",\n",
       " \"['GTDB:Mycobacterium noviomagense']\",\n",
       " \"['GTDB:Mycobacterium phocaicum']\",\n",
       " \"['GTDB:Mycobacterium saopaulense']\",\n",
       " \"['GTDB:Mycobacterium seoulense']\",\n",
       " \"['GTDB:Mycobacterium septicum']\",\n",
       " \"['GTDB:Mycobacterium shigaense']\",\n",
       " \"['GTDB:Mycobacterium sp000328565']\",\n",
       " \"['GTDB:Mycobacterium stomatepiae']\",\n",
       " \"['GTDB:Mycobacterium tuberculosis']\",\n",
       " \"['GTDB:Mycoplasmoides amphoriforme']\",\n",
       " \"['GTDB:Mycoplasmoides genitalium']\",\n",
       " \"['GTDB:Mycoplasmoides pneumoniae']\",\n",
       " \"['GTDB:Mycoplasmopsis fermentans']\",\n",
       " \"['GTDB:Neisseria brasiliensis']\",\n",
       " \"['GTDB:Neisseria flavescens_A']\",\n",
       " \"['GTDB:Neisseria flavescens_B']\",\n",
       " \"['GTDB:Neisseria flavescens_C']\",\n",
       " \"['GTDB:Neisseria gonorrhoeae']\",\n",
       " \"['GTDB:Neisseria meningitidis']\",\n",
       " \"['GTDB:Neisseria perflava']\",\n",
       " \"['GTDB:Neorickettsia sennetsu']\",\n",
       " \"['GTDB:Nocardia asiatica']\",\n",
       " \"['GTDB:Nocardia brasiliensis_B']\",\n",
       " \"['GTDB:Nocardia cyriacigeorgica_B']\",\n",
       " \"['GTDB:Nocardia exalbida']\",\n",
       " \"['GTDB:Nocardia farcinica']\",\n",
       " \"['GTDB:Nocardia terpenica']\",\n",
       " \"['GTDB:Nocardia wallacei']\",\n",
       " \"['GTDB:Nocardiopsis dassonvillei']\",\n",
       " \"['GTDB:Ochrobactrum anthropi']\",\n",
       " \"['GTDB:Ochrobactrum intermedium']\",\n",
       " \"['GTDB:Odoribacter splanchnicus']\",\n",
       " \"['GTDB:Olsenella uli']\",\n",
       " \"['GTDB:Oribacterium parvum']\",\n",
       " \"['GTDB:Orientia tsutsugamushi']\",\n",
       " \"['GTDB:Pandoraea apista']\",\n",
       " \"['GTDB:Pandoraea fibrosis']\",\n",
       " \"['GTDB:Pandoraea pnomenusa']\",\n",
       " \"['GTDB:Pannonibacter indicus']\",\n",
       " \"['GTDB:Pantoea ananatis']\",\n",
       " \"['GTDB:Parabacteroides distasonis']\",\n",
       " \"['GTDB:Paraclostridium sordellii']\",\n",
       " \"['GTDB:Parascardovia denticolens']\",\n",
       " \"['GTDB:Parasutterella excrementihominis']\",\n",
       " \"['GTDB:Parvimonas micra']\",\n",
       " \"['GTDB:Pasteurella multocida']\",\n",
       " \"['GTDB:Pauljensenia odontolytica_A']\",\n",
       " \"['GTDB:Pauljensenia odontolytica_C']\",\n",
       " \"['GTDB:Peptidiphaga gingivicola']\",\n",
       " \"['GTDB:Phocaeicola massiliensis']\",\n",
       " \"['GTDB:Phocaeicola vulgatus']\",\n",
       " \"['GTDB:Photorhabdus asymbiotica']\",\n",
       " \"['GTDB:Plesiomonas shigelloides']\",\n",
       " \"['GTDB:Porphyromonas gingivalis']\",\n",
       " \"['GTDB:Prevotella bivia']\",\n",
       " \"['GTDB:Prevotella intermedia']\",\n",
       " \"['GTDB:Proteus mirabilis']\",\n",
       " \"['GTDB:Proteus penneri']\",\n",
       " \"['GTDB:Providencia alcalifaciens']\",\n",
       " \"['GTDB:Providencia huaxiensis']\",\n",
       " \"['GTDB:Providencia rettgeri_D']\",\n",
       " \"['GTDB:Providencia_C stuartii']\",\n",
       " \"['GTDB:Pseudoflavonifractor capillosus']\",\n",
       " \"['GTDB:Pseudoleptotrichia goodfellowii']\",\n",
       " \"['GTDB:Pseudomonas aeruginosa']\",\n",
       " \"['GTDB:Pseudomonas aeruginosa_A']\",\n",
       " \"['GTDB:Pseudomonas_E fluorescens_AP']\",\n",
       " \"['GTDB:Pseudomonas_E haemolytica']\",\n",
       " \"['GTDB:Pseudomonas_E iridis']\",\n",
       " \"['GTDB:Pseudomonas_E lactis']\",\n",
       " \"['GTDB:Pseudomonas_E lundensis']\",\n",
       " \"['GTDB:Pseudomonas_E lurida']\",\n",
       " \"['GTDB:Pseudomonas_E paracarnis']\",\n",
       " \"['GTDB:Pseudomonas_E parakoreensis']\",\n",
       " \"['GTDB:Pseudomonas_E protegens']\",\n",
       " \"['GTDB:Pseudomonas_E putida']\",\n",
       " \"['GTDB:Pseudomonas_E rhodesiae']\",\n",
       " \"['GTDB:Rhodococcus equi']\",\n",
       " \"['GTDB:Rhodococcus gordoniae']\",\n",
       " \"['GTDB:Rickettsia akari']\",\n",
       " \"['GTDB:Rickettsia felis']\",\n",
       " \"['GTDB:Rickettsia prowazekii']\",\n",
       " \"['GTDB:Rickettsia rhipicephali']\",\n",
       " \"['GTDB:Rickettsia rickettsii']\",\n",
       " \"['GTDB:Rickettsia typhi']\",\n",
       " \"['GTDB:Rothia aeria']\",\n",
       " \"['GTDB:Rothia mucilaginosa_B']\",\n",
       " \"['GTDB:Rothia sp001808955']\",\n",
       " \"['GTDB:Saccharomonospora viridis']\",\n",
       " \"['GTDB:Salmonella arizonae']\",\n",
       " \"['GTDB:Salmonella enterica']\",\n",
       " \"['GTDB:Sarcina perfringens']\",\n",
       " \"['GTDB:Serratia marcescens_K']\",\n",
       " \"['GTDB:Shewanella xiamenensis']\",\n",
       " \"['GTDB:Simkania negevensis']\",\n",
       " \"['GTDB:Sneathia vaginalis']\",\n",
       " \"['GTDB:Staphylococcus arlettae']\",\n",
       " \"['GTDB:Staphylococcus aureus']\",\n",
       " \"['GTDB:Staphylococcus capitis']\",\n",
       " \"['GTDB:Staphylococcus epidermidis']\",\n",
       " \"['GTDB:Staphylococcus haemolyticus']\",\n",
       " \"['GTDB:Staphylococcus lugdunensis']\",\n",
       " \"['GTDB:Staphylococcus massiliensis']\",\n",
       " \"['GTDB:Staphylococcus saprophyticus']\",\n",
       " \"['GTDB:Staphylococcus warneri']\",\n",
       " \"['GTDB:Stenotrophomonas maltophilia']\",\n",
       " \"['GTDB:Stenotrophomonas maltophilia_F']\",\n",
       " \"['GTDB:Streptobacillus moniliformis']\",\n",
       " \"['GTDB:Streptococcus agalactiae']\",\n",
       " \"['GTDB:Streptococcus anginosus']\",\n",
       " \"['GTDB:Streptococcus anginosus_C']\",\n",
       " \"['GTDB:Streptococcus constellatus']\",\n",
       " \"['GTDB:Streptococcus dysgalactiae']\",\n",
       " \"['GTDB:Streptococcus equi']\",\n",
       " \"['GTDB:Streptococcus gallolyticus']\",\n",
       " \"['GTDB:Streptococcus gordonii']\",\n",
       " \"['GTDB:Streptococcus ilei']\",\n",
       " \"['GTDB:Streptococcus infantis_I']\",\n",
       " \"['GTDB:Streptococcus iniae']\",\n",
       " \"['GTDB:Streptococcus intermedius']\",\n",
       " \"['GTDB:Streptococcus mitis_M']\",\n",
       " \"['GTDB:Streptococcus mitis_O']\",\n",
       " \"['GTDB:Streptococcus mitis_Q']\",\n",
       " \"['GTDB:Streptococcus mutans']\",\n",
       " \"['GTDB:Streptococcus oralis_BC']\",\n",
       " \"['GTDB:Streptococcus oriscaviae']\",\n",
       " \"['GTDB:Streptococcus parasanguinis']\",\n",
       " \"['GTDB:Streptococcus pneumoniae']\",\n",
       " \"['GTDB:Streptococcus pseudopneumoniae']\",\n",
       " \"['GTDB:Streptococcus pyogenes']\",\n",
       " \"['GTDB:Streptococcus salivarius']\",\n",
       " \"['GTDB:Streptococcus sanguinis']\",\n",
       " \"['GTDB:Streptococcus sanguinis_H']\",\n",
       " \"['GTDB:Streptococcus sobrinus']\",\n",
       " \"['GTDB:Streptococcus sp002355895']\",\n",
       " \"['GTDB:Streptococcus suis']\",\n",
       " \"['GTDB:Streptococcus symci']\",\n",
       " \"['GTDB:Streptococcus urinalis']\",\n",
       " \"['GTDB:Tannerella forsythia']\",\n",
       " \"['GTDB:Thomasclavelia spiroformis']\",\n",
       " \"['GTDB:Treponema pallidum']\",\n",
       " \"['GTDB:Treponema sp000775995']\",\n",
       " \"['GTDB:Treponema sp010365865']\",\n",
       " \"['GTDB:Treponema_B denticola']\",\n",
       " \"['GTDB:Treponema_B putidum']\",\n",
       " \"['GTDB:Tropheryma whipplei']\",\n",
       " \"['GTDB:Tsukamurella paurometabola']\",\n",
       " \"['GTDB:Tsukamurella tyrosinosolvens']\",\n",
       " \"['GTDB:Ureaplasma parvum']\",\n",
       " \"['GTDB:Ureaplasma urealyticum']\",\n",
       " \"['GTDB:Vibrio alginolyticus']\",\n",
       " \"['GTDB:Vibrio cholerae']\",\n",
       " \"['GTDB:Vibrio furnissii']\",\n",
       " \"['GTDB:Vibrio harveyi']\",\n",
       " \"['GTDB:Vibrio mimicus']\",\n",
       " \"['GTDB:Vibrio owensii']\",\n",
       " \"['GTDB:Vibrio parahaemolyticus']\",\n",
       " \"['GTDB:Vibrio vulnificus']\",\n",
       " \"['GTDB:Yersinia enterocolitica']\",\n",
       " \"['GTDB:Yersinia kristensenii']\",\n",
       " \"['GTDB:Yersinia pestis']\"}"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtdbid_df.loc[gtdbid_df['gtdb_id'].isin(all_potential_pathogenic_strain),:].reset_index(drop=True).to_csv(os.path.join(args.output_dir, 'all_potential_pathogenic_strain.tsv'), sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "edge_info.to_csv(os.path.join(args.output_dir, 'edge_info.tsv'), sep='\\t', index=False) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_seed(100)\n",
    "species1_list = list(species1)\n",
    "random.shuffle(species1_list)\n",
    "hold_out_species = species1_list[:int(len(species1_list)*0.1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "pathogen_info.loc[pathogen_info['parent_list'].isin(hold_out_species),:].to_csv(os.path.join(args.output_dir, 'hold_out_pathogen_info.tsv'), sep='\\t', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mkg_model_training",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
