## Add VFDB and AMR to the graph

More refs:

1. Benchmark of different VF database: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3411817/
2. PATRIC curated VFDB for infectious disease: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4287947/
3. AMR data:
   1. PATRIC has an AMR phenotype table generated by laboratory methods. This table contains records like `32002.4 tiamulin` but NO AMR gene sequences. These records may serve as gold standard training data. 
   2. PATRIC also curated AMR genes mainly from CARD
   3. PATRIC has an AMR genome sets by antibiotic-resistant or suseptible category. But not what we need here.



Data location:

```
### Local at CPU server:
/scratch2/sml6467/temp/VFDB
/scratch2/sml6467/temp/AMR

### in our graph folder
/data/shared_data/metagenomics_graph/pathogen_database/VFDB
/data/shared_data/metagenomics_graph/pathogen_database/AMR
```











### VFDB

---

Download data from [VFDB ](http://www.mgc.ac.cn/VFs/download.htm)or from the curated VFDB in PATRIC FTB site (FTP -> specialty genes -> refDB). Record number:

| PATRIC curated | PATRIC_VFDB | VFDB_core | VFDB_full |
| -------------- | ----------- | --------- | --------- |
| 1570           | 2595        | **4236**  | 27982     |

It seems like the PATRIC VFDB is not up to date and it specifically focuses on some diseases (see ref2 above). So I think we should use the VFDB_core data.



There are 6 files:

1. an excel explaining each VF 
2. a comparative table for VF within a bacteria group (we don't need this)
3. DNA + **Protein** sequence for **core datasets** (representative genes associated with experimentally verified VFs only)
4. DNA + **Protein** sequence for **full datasets** (all genes related to known and predicted VFs)



Use these codes to download files:

```
wget http://www.mgc.ac.cn/VFs/Down/VFDB_setA_nt.fas.gz
wget http://www.mgc.ac.cn/VFs/Down/VFDB_setA_pro.fas.gz
wget http://www.mgc.ac.cn/VFs/Down/VFDB_setB_nt.fas.gz
wget http://www.mgc.ac.cn/VFs/Down/VFDB_setB_pro.fas.gz

# manually transfer the excel to csv file
# wget http://www.mgc.ac.cn/VFs/Down/VFs.xls.gz
```

File information:

| Name              | Contents                                             |
| ----------------- | ---------------------------------------------------- |
| VFs.csv           | Information file with VFID, name, category, function |
| VFDB_setA_nt.fas  | DNA of sub-unit of each VF for core data             |
| VFDB_setA_pro.fas | Protein of sub-unit of each VF for core data         |
| VFDB_setB_nt.fas  | DNA of sub-unit of each VF for full data             |
| VFDB_setB_pro.fas | Protein of sub-unit of each VF for full data         |



#### Final files for graph building:

Structure in graph: VFID <-> several sub-unites <-> genome

1. VFDB_setA_pro.fas: protein sequences for core datasets. 
   1. VF id in the middle paranthesis, e.g. Phospholipase C (**VF0470**) - Exotoxin (VFC0235)
   2. every record is a **SUB-UNIT** of an ID, NOT the whole ID
2. VFs.csv: metadate for the VF file



### AMR

---

#### The PATRIC AMR phenotype table:

```
/data/shared_data/metagenomics_graph/pathogen_database/BV-BRC/raw_data/PATRIC_genomes/PATRIC_genomes_AMR.txt
```

col1: genome id

col4: AMR phenotype



#### AMR gene sequences

Download at [CARD](https://card.mcmaster.ca/download), I'll use the "protein_homolog_model" data

```
wget https://card.mcmaster.ca/download/0/broadstreet-v3.2.7.tar.bz2
tar -xjf broadstreet-v3.2.7.tar.bz2
```



| PATRIC curated | CARD protein_homolog_model | CARD protein_knockout_model | CARD  protein_overexpression_model | CARD  protein_variant_model |
| -------------- | -------------------------- | --------------------------- | ---------------------------------- | --------------------------- |
| 2277           | 4775                       | 18                          | 13                                 | 178                         |

Explanation of models: 

the "protein homolog" model type contains sequences of antimicrobial resistance genes that do not include mutation as a determinant of resistance - these data are appropriate for BLAST analysis of metagenomic data or searches excluding secondary screening for resistance mutations. In contrast, the "protein variant" model includes reference wild type sequences used for mapping SNPs conferring antimicrobial resistance - without secondary mutation screening, analyses using these data will include false positives for antibiotic resistant gene variants or mutants.



#### Final files:

1. protein_fasta_protein_homolog_model.fasta: protein sequences of AMR genes
   1. record structure: `>gb|ACT97415.1|ARO:3002999|CblA-1 [mixed culture bacterium AX_gF3SD01_15]`
      1. ACT97415.1: CARD id
      2. ARO:3002999: Accession in GenBank
      3. CblA-1: CARD short name (no other meaning, just for programming purpose)
   2. inside the file they use full name of the pathogens, so don't need short name file for species
2. CARD-Download-README.txt: full document
3. (not sure if needed) aro_categories.tsv: map records to categories by ARO
4. aro_index.tsv: the metadata where you can map ARO id to more information such as gene family, drug class, resistant mechanism
5. shortname_antibiotics.tsv: maps drug class short names to full name here


